[#12] 경로검색 페이지 구현

hyuns66 opened the pull request:
## #️⃣연관된 이슈

- #12 

## 📝작업 내용

- 일단 UI작업만 했습니다
- 애니메이션의 경우 모호한 부분이 있어 디자이너님께 여쭤보고 추후 구현완료할 예정입니다
- 디자인시스템에 약간의 변경사항이 생겼습니다
- TabRowMinTab0dp 의 경우 이전 디자인에서 작은 사이즈의 탭 레이아웃을 만들기 위한 컴포넌트인데 디자인이 변경되면서 사용하지 않게 되었지만 혹시몰라 코드는 남겨두었습니다. (삭제할까 말까 추천좀용)
- 

## PR 발행 전 체크 리스트

- [x] 발행자 확인
- [x] 프로젝트 설정 확인
- [x] 라벨 확인
- [x] 코드 린트 확인

## 스크린샷 (선택)
![image](https://github.com/user-attachments/assets/d0e67487-8430-4bd9-be24-a2b21c1c270b)
![image](https://github.com/user-attachments/assets/341692a8-f4a3-4dfe-b6c0-ccfe40432e08)

## 💬리뷰 요구사항(선택)

- 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
  - @cyjadela 디자인시스템쪽에 색상 몇가지 추가됐는데 확인 함만 해주세요!

hyuns66 committed:
feat#12 : 상단 컴포넌트 추가

- 앱  바 관련 컴포넌트
- 검색창 컴포넌트

추가했습니다

hyuns66 committed:
feat#12 : 상단 컴포넌트 추가

- 앱  바 관련 컴포넌트
- 검색창 컴포넌트

추가했습니다

hyuns66 committed:
feat/#12 : 탭 레이아웃 커스텀 컴포넌트 추가

Material에서 제공하는 탭 레이아웃은 minwidth가 지역변수로 정해져있어 탭
사이즈를 커스텀할 수 없는 문제가 있었습니다.

따라서 minWidth = 0dp인 커스텀 TabLayout 컴포넌트를 만들었습니다.

hyuns66 committed:
feat/#12 : 탭 레이아웃 커스텀 컴포넌트 추가

Material에서 제공하는 탭 레이아웃은 minwidth가 지역변수로 정해져있어 탭
사이즈를 커스텀할 수 없는 문제가 있었습니다.

따라서 minWidth = 0dp인 커스텀 TabLayout 컴포넌트를 만들었습니다.

Commit:
feat/#12 : typography 추가

course 화면에서 heading4 medium 20dp 폰트가 추가되었습니다
디자인 시스템 등록해놓았고 새로운 폰트 담당자는 사랑님입니다!

# Conflicts:
#	app/src/main/res/values/strings.xml

Commit:
feat/#12 : typography 추가

course 화면에서 heading4 medium 20dp 폰트가 추가되었습니다
디자인 시스템 등록해놓았고 새로운 폰트 담당자는 사랑님입니다!

cyjadela committed:
feat/#21 : FCM google-services 파일 추가

cyjadela committed:
feat/#21 : Firebase sdk 의존성 추가

cyjadela committed:
feat/#21 : FirebaseMessagingService 구현

cyjadela committed:
feat/#21 : FCM 토큰 추출

cyjadela committed:
chore/#21 : ktlint 적용

cyjadela committed:
chore/#21 : FCM google-services 파일 삭제

cyjadela committed:
docs/#21 : FCM google-services 파일 gitignore 제거

cyjadela committed:
feat/#21 : FCM google-services 추가

t1nm1ksun committed:
chore/#21 : google-service 플러그인 중복 선언 주석처리

t1nm1ksun committed:
Merge branch 'feat/fcm#21' of https://github.com/depromeet/16th-team6-Android into feat/fcm#21

hyuns66 committed:
feat#12 : 디자인 시스템 색상코드 추가

디자인 시스템에 Non symantic color grey 계열이 추가되어서 코드에 추가했습니다

hyuns66 committed:
feat#12 : 디자인 시스템 색상코드 추가

디자인 시스템에 Non symantic color grey 계열이 추가되어서 코드에 추가했습니다

hyuns66 committed:
feat#12 : courseItem UI 완성

재사용성 없는 컴포넌트는 굳이 파일로 분리 안했습니다.

hyuns66 committed:
feat#12 : courseItem UI 완성

재사용성 없는 컴포넌트는 굳이 파일로 분리 안했습니다.

hyuns66 committed:
feat#12 : CourseScreen 완성 및 경로상세보기 토글 애니메이션 추가

- 현재 피그마 디자인상 경로상세보기를 펼쳤을 때 애니메이션이 어떻게 들어가야할지 모호하여 펼쳤을 때의 컨텐츠 구현은 화요일 디자이너님 회의 때 여쭤보고 완성하겠습니다

hyuns66 committed:
feat#12 : CourseScreen 완성 및 경로상세보기 토글 애니메이션 추가

- 현재 피그마 디자인상 경로상세보기를 펼쳤을 때 애니메이션이 어떻게 들어가야할지 모호하여 펼쳤을 때의 컨텐츠 구현은 화요일 디자이너님 회의 때 여쭤보고 완성하겠습니다

hyuns66 committed:
chore#12 : linting

hyuns66 committed:
chore#12 : linting

cyjadela committed:
chore/#21 : google-services 파일 제거

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/CourseAppBar.kt:
```diff
@@ -0,0 +1,49 @@
+        horizontalArrangement = Arrangement.SpaceBetween,
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Icon(
+            painter = painterResource(R.drawable.ic_all_arrow_left_white),
+            contentDescription = "appbar back",
+            tint = defaultTeam6Colors.white,
+            modifier = Modifier.size(24.dp)
+        )
+
+        Icon(
+            painter = painterResource(R.drawable.ic_all_home_white),
+            contentDescription = "appbar home",
+            tint = defaultTeam6Colors.white,
+            modifier = Modifier.size(24.dp)
+        )
```
노션에도 적어두었는데 svg를 이용한 벡터 이미지를 사용할 때는 painter보다는 imageVectorResource를 사용하는게 immutable함을 보장하여 불필요한 Recomposition을 방지할 수 있습니다! (png같은 이미지는 안되긴해욥)

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/DestinationSearchBar.kt:
```diff
@@ -0,0 +1,72 @@
+    Row(
+        modifier = modifier
+            .fillMaxWidth()
+            .background(
+                color = defaultTeam6Colors.greyElevatedCard,
+                shape = RoundedCornerShape(12.dp)
+            )
+            .padding(vertical = 12.dp),
```
modifeir 확장함수 파일 보시면 한번에 할수있는 함수 만들어져있습니다람쥐~

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/DestinationSearchBar.kt:
```diff
@@ -0,0 +1,72 @@
+                color = defaultTeam6Colors.greyElevatedCard,
+                shape = RoundedCornerShape(12.dp)
+            )
+            .padding(vertical = 12.dp),
+        horizontalArrangement = Arrangement.SpaceBetween,
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        // TODO : typography 디자인 시스템 적용
+        Text(
+            text = startingPoint,
+            fontSize = 16.sp,
+            color = defaultTeam6Colors.greySecondaryLabel,
+            modifier = Modifier
+                .padding(start = 62.dp)
+        )
```
<img width="400" alt="image" src="https://github.com/user-attachments/assets/9403b0c6-cba3-42e9-84b9-7dbca4ebf236" />

위 사진의 컴포넌트를 작업하신거 맞으실까요?
제가 잘못알고 있을 수도 있어서요...ㅎ
맞으시다면 패딩값 피그마 디자인 값 확인 부탁드립니다!

hyuns66 replied:
아 저부분은 예전 디자인으로 작업되어있는건데 디자이너님께 물어볼 부분이 있어서 추후 수정할거에요!

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/LastTransportInfoItem.kt:
```diff
@@ -0,0 +1,249 @@
+    lastTransportInfo: LastTransportInfo,
+    modifier: Modifier = Modifier
+) {
+    Column(
+        modifier = modifier
+            .clip(RoundedCornerShape(12.dp))
+            .background(defaultTeam6Colors.greyCard)
+            .padding(16.dp)
```
<img width="549" alt="image" src="https://github.com/user-attachments/assets/15da2501-82d9-46ac-ac67-0abced14431a" />

위에도 남겨드렸지만 요런식으로 만들어준 확장함수 쓰시면 편할것 같아요~

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/LastTransportInfoItem.kt:
```diff
@@ -0,0 +1,249 @@
+            }
+        }
+
+        Spacer(
+            modifier = Modifier
+                .fillMaxWidth()
+                .height(10.dp)
+        )
```
Spacer에 색을 따로 지정할게 아니라면 fillMaxWidth 안해주셔도 될 것 같아요!
아니면 다른 의도가 있으실까요?!

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/LastTransportInfoItem.kt:
```diff
@@ -0,0 +1,249 @@
+fun SetNotificationButton(
+    modifier: Modifier = Modifier
+) {
+    Row(
+        modifier = Modifier
+            .clip(RoundedCornerShape(8.dp))
+            .background(color = Color(0x1F8AF265))
+            .padding(vertical = 13.dp, horizontal = 28.dp)
+            .fillMaxWidth(),
+        horizontalArrangement = Arrangement.Center,
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Image(
+            modifier = Modifier
+                .size(16.dp),
+            painter = painterResource(R.drawable.ic_all_alarm_clock_green),
```
요오기도 위에 코멘트 확인해주세욥!

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/LastTransportInfoItem.kt:
```diff
@@ -0,0 +1,249 @@
+    isDeparture: Boolean,
+    modifier: Modifier = Modifier
+) {
+    Box(
+        modifier = Modifier
+            .clip(RoundedCornerShape(20.dp))
+            .background(defaultTeam6Colors.systemGrey5)
+    ) {
+        val color = if (isDeparture) {
+            defaultTeam6Colors.primaryMain
+        } else {
+            defaultTeam6Colors.white
+        }
+        Text(
+            modifier = Modifier
+                .padding(vertical = 4.dp, horizontal = 8.dp),
+            color = color,
+            text = stringResource(R.string.last_transport_info_remaining_time, hour, minute),
+            style = defaultTeam6Typography.bodySemiBold12
+        )
```
Text의 modifier를 잘 활용하면 Box로 감싸지 않고 불필요한 layout depth를 줄일 수 있을 것 같아요!

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/LastTransportInfoItem.kt:
```diff
@@ -0,0 +1,249 @@
+    }
+}
+
+@Preview(name = "more than 1 hour", showBackground = true, backgroundColor = android.graphics.Color.BLACK.toLong())
```
오 Preview속성에 이렇게 지정해주면 Preview에서 부모 Layout을 색상을 지정하주고 감쌀필요가 없네요?!
하나 배워갑니다~

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/TabRowMinTab0dp.kt:
```diff
@@ -0,0 +1,245 @@
+        HorizontalDivider()
+    },
+    scrollState: ScrollState = rememberScrollState(),
+    tabs: @Composable () -> Unit
```
필수인자는 optional parameter보다 위로 올려주시면 좋을 것 같습니다!

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/TabRowMinTab0dp.kt:
```diff
@@ -0,0 +1,245 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+@Composable
+fun TabRowMinTab0dp(
```
요 친구는 지도에 마커찍는건가요?

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/TransportCourseInfo.kt:
```diff
@@ -0,0 +1,131 @@
+    modifier: Modifier = Modifier
+) {
+    val context = LocalContext.current
+    var expanded by remember { mutableStateOf(false) }
```
```suggestion
    var expanded by rememberSaveable { mutableStateOf(false) }
```

rememberSaveable 을 사용하여 관리하면 Activity 재생성시에도 상태가 유지되기 때문에 안정성 측면에서 좋을 것 같아요!

hyuns66 replied:
remember에서는 상태가 RAM 메모리에 저장되는 반면 rememberSaveable은 Bundle에 저장하는거로 알고있습니다!
애니메이션같은 경우는 데이터 유지보단 가볍고 빠른 속도가 더 중요하다 판단해서 remember를 사용했어요

반면에 텍스트 필드나 체크박스 같은 뷰의 경우는 데이터 유지가 더 중요하기 때문에 rememberaSaveable를 쓰면 좋을것 같습니다 :)

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/TransportTabMenu.kt:
```diff
@@ -0,0 +1,94 @@
+    availableCourses: List<LastTransportInfo>,
+    modifier: Modifier = Modifier
+) {
+    val tabItems = listOf("전체", "버스", "지하철")
```
```suggestion
    val tabItems = immutableListOf("전체", "버스", "지하철")
```

바뀌는 항목들이 아니라면? 저희 스터디 때 이야기한 내용처럼 immutable한 list로 설정하는게 Stable해질 것 같아요!

hyuns66 replied:
앗 그러게요 
저건 그냥 리소스파일 배열로 뺄게요!

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/presentation/ui/course/component/TransportTabRow.kt:
```diff
@@ -0,0 +1,155 @@
+        value = currentTabPosition
+    }
+) {
+    val currentTabWidth by animateDpAsState(
+        targetValue = tabWidth,
+        animationSpec = tween(durationMillis = 250, easing = FastOutSlowInEasing)
+    )
+    val indicatorOffset by animateDpAsState(
+        // 1dp -> 인디케이터가 정중앙에 오도록 보정
+//        targetValue = ((currentTabPosition.left + currentTabPosition.right - tabWidth) / 2) + 1.dp,
+//        animationSpec = tween(durationMillis = 250, easing = FastOutSlowInEasing)
+        targetValue = currentTabPosition.left,
+        animationSpec = tween(durationMillis = 250, easing = FastOutSlowInEasing)
+    )
```
250도 따로 네이밍해서 상수로 빼주시면 좋을 것 같아요!

t1nm1ksun reviewed (commented):
정처기랑 코테준비하시느라 바쁘셨을텐데 고생하셨습니다...!!!

cyjadela committed:
fix/#21 : conflict 해결

cyjadela committed:
Merge branch 'develop' into feat/fcm#21

cyjadela committed:
chore/#21 : ci actions/cache version 변경

cyjadela committed:
Merge pull request #22 from depromeet/feat/fcm#21

[#21] FCM 알림 구현

hyuns66 committed:
Update app/src/main/java/com/depromeet/team6/presentation/ui/course/CourseScreen.kt

Co-authored-by: 신민석 <122257945+t1nm1ksun@users.noreply.github.com>

hyuns66 committed:
fix#12 : 코드리뷰 반영 - bitmap -> vectorResource

hyuns66 committed:
fix#12 : 코드 리뷰 반영했습니다.

- arrays.xml파일로 탭 메뉴 배열 이동
- 필수 인자 순서 변경
- RemainingTimeHHmm 컴포넌트 레이아웃 depth 감소

hyuns66 committed:
Merge branch 'feat/course-search#12' of https://github.com/depromeet/16th-team6-Android into feat/course-search#12

hyuns66 committed:
chore#12 : linting

jieeeunnn reviewed (approved):
민석님이 코드 리뷰를 넘 꼼꼼하게 남겨주셔서 더 남길 게 없네요,, 👍🏻
근데 윤진님 코드 rebase 하신 건가요? 커밋에 윤진님 기록이 있길래 여쭤봅니다!
(rebase를 많이 안써봐서 .. 👀)

hyuns66 commented:
> 민석님이 코드 리뷰를 넘 꼼꼼하게 남겨주셔서 더 남길 게 없네요,, 👍🏻 근데 윤진님 코드 rebase 하신 건가요? 커밋에 윤진님 기록이 있길래 여쭤봅니다! (rebase를 많이 안써봐서 .. 👀)

ㅋㅋㅋ 컴 바꾸고 깃 새로설치했더니 설정이 초기화돼서 pull하려고 했더니 리베이스할래? 머지할래? 물어보더라구요
저도 리베이스 안해봐서 한번 해본건데 좀 잘못만져서 히스토리 꼬여가지고 그냥 머지하는거로 다시 바꿨습니다....ㅋㅋㅋㅋ
윤진님 기록은 아마 컨플릭트 해결하느라고 develop 브랜치 머지하는 과정에서 생긴거 같아요

hyuns66 committed:
fix#12 : typedArray autoclosable 타입 활용시 minsdk 지원가능하도록 수정

원래는 .usd 블럭을 활용해서 autoclosable 객체의 특성을 활용하고자 했지만
프로젝트의 minsdk는 26으로 지정되어있어 해당기능이 지원되지 않기에
임의로 recycle()를 호출하도록 수정했습니다.

hyuns66 committed:
feat#12 : course search 컴포넌트 변경

hyuns66 committed:
chore#12 : 경로 검색, 상세 페이지 분기하기 위해 course -> course_search 패키지명 변경

jieeeunnn commented:
> ㅋㅋㅋ 컴 바꾸고 깃 새로설치했더니 설정이 초기화돼서 pull하려고 했더니 리베이스할래? 머지할래? 물어보더라구요 저도 리베이스 안해봐서 한번 해본건데 좀 잘못만져서 히스토리 꼬여가지고 그냥 머지하는거로 다시 바꿨습니다....ㅋㅋㅋㅋ 윤진님 기록은 아마 컨플릭트 해결하느라고 develop 브랜치 머지하는 과정에서 생긴거 같아요

아하 ㅋㅋㅋㅋ 넵 확인했습니다 ㅎㅎ

cyjadela reviewed (approved):
고생하셨습니다 ~ !!!!!! 제 이름 있어서 저 뭐 잘못한 줄 알고 깜짝 놀랐네요 ,,, ㅎㅎ
디자인 시스템 확인했습니다 ! 저도 gray 만들어놨긴 한데 제꺼 나중에 삭제하고 합칠게요 ~ !

hyuns66 committed:
feat#12 : 경로 상세화면 업데이트

- 디자이너님과 논의 후 1차 MVP에서는 애니메이션 없이 펼치는거로 구현했습니다.

hyuns66 committed:
chore#12 : linting

hyuns66 committed:
fix#12 : 경로상세 정보 펼칠 시 코너라운딩값 변경되도록 수정

hyuns66 committed:
feat#12 : 경로 검색창에서 길이초과시 elipsize 처리

hyuns66 committed:
feat#12 : add contract and viewModel

- AnimatedSnackbar의 경우 추후 수정예정입니다.
